homeassistant/components/kermi/__init__.py

```diff
@@ -0,0 +1,107 @@
+"""Kermi heatpump integration for Home Assistant."""
+
+from datetime import timedelta
+from functools import partial
+import logging
+
+from pymodbus.client import AsyncModbusTcpClient
+from pymodbus.exceptions import ConnectionException, ModbusIOException
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_HOST
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DOMAIN, MODBUS_REGISTERS
+
+# Create a logger
+logger = logging.getLogger(__name__)
+
+PLATFORMS = ["water_heater"]
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up kermi from a config entry."""
+    hass.data.setdefault(DOMAIN, {})
+
+    # Create a ModbusTcpClient
+    client = AsyncModbusTcpClient(host=entry.data[CONF_HOST], port=502)
+
+    # Connect to the client
+    try:
+        await client.connect()
+    except ConnectionException as err:
+        raise ConfigEntryNotReady from err
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        logger,
+        name="kermi",
+        update_method=partial(
+            async_update_data, client, entry.data["water_heater_device_address"]
+        ),
+        update_interval=timedelta(minutes=1),
+    )
+
+    hass.data[DOMAIN][entry.entry_id] = {
+        "client": client,
+        "coordinator": coordinator,
+    }
+
+    # Forward the setup to the water heater platform
+    await hass.config_entries.async_forward_entry_setups(entry, ["water_heater"])
+
+    try:
+        # Fetch initial data
+        await coordinator.async_config_entry_first_refresh()
+    except ModbusIOException as e:
+        logger.error("Modbus IO exception: %s", e)
+        return False
+
+    if not coordinator.data:
+        return False
+
+    return True
+
+
+async def async_update_data(client, device_address):
```

Comment:
for cleanlyness I'd prefer you'd create your own Coordinator class and put this in a `coordinator.py`